include/orb/GroundPropagator.h

```diff
@@ -0,0 +1,173 @@
+     */
+    Orbit to_catch_up;
+
+  public:
+    /**
+     * Construct GroundPropagator.
+     *
+     * grav calls: 0
+     */
+    GroundPropagator(){}
+
+    /** Input the newest Orbit sent from ground.
+     * 
+     * grav calls: 0
+     * @param[in] ground_data: Orbit sent from ground.
+     * @param[in] gps_time_ns: Time to propagate to (ns).
+     * @param[in] earth_rate_ecef: The earth's angular rate in ecef frame ignored for orbits already propagating(rad/s).
+     */
+    void input(const Orbit& ground_data, const uint64_t& gps_time_ns, const lin::Vector3d& earth_rate_ecef){
+        //add to queue
+        if (ground_data.valid()){
+            if(!current.valid()){
+                current= ground_data;
+            } else if(!catching_up.valid()){
+                catching_up= ground_data;
+            } else{
+                to_catch_up= ground_data;
+                //Assuming the new ground_data is closer to the current time 
+                //than previous ground data, over writing this shouldn't be losing any important 
+                //information. 
+                //If that assumion is wrong, then the estimator isn't optimal, but it still works.
+            }
+        }
+        //start propagators
+        current.startpropagating(gps_time_ns,earth_rate_ecef);
+        catching_up.startpropagating(gps_time_ns,earth_rate_ecef);
+        to_catch_up.startpropagating(gps_time_ns,earth_rate_ecef);
+        //sort stuff
+        if (to_catch_up.valid() && to_catch_up.numgravcallsleft()<=catching_up.numgravcallsleft()){
+            catching_up= to_catch_up;
+            to_catch_up= Orbit();
+        }
+        if (catching_up.valid() && catching_up.numgravcallsleft()<=current.numgravcallsleft()){
+            current= catching_up;
+            catching_up= to_catch_up;
+            to_catch_up= Orbit();
+        }
+    }
+
+    /** Do one grav call to move the propagator towards the current time if needed.
+     * The Orbit closest to finish propagating is propagated first.
+     * 
+     * grav calls: 1 or 0 if totalnumgravcallsleft() == 0 
+     */
+    void onegravcall(){
+        if(current.numgravcallsleft()){
+            current.onegravcall();
+        } else { 
+            catching_up.onegravcall();
+        }
+        //sort stuff
+        if (catching_up.valid() && catching_up.numgravcallsleft()<=current.numgravcallsleft()){
+            current= catching_up;
+            catching_up= to_catch_up;
+            to_catch_up= Orbit();
+        }
+    }
+
+    /** Return the total number of grav calls need to finish propagating all the orbits.
+     *
+     * grav calls: 0
+     */
+    int totalnumgravcallsleft(){
+        return current.numgravcallsleft()+catching_up.numgravcallsleft()+to_catch_up.numgravcallsleft();
+    }
+
+    /** Returns a reference to the best estimate of the Orbit.
+     * 
+     * grav calls: 0
+     */
+    Orbit bestestimate() const{
+        return current;
+    }
```

Comment:
How exactly is this going to integrate with the orbit filter, or are we just going to propagate and not filter orbits received from the ground (which would totally be fine)?

Reply:
Yeah, the main orbit estimator is going to have a ground propagator for both the self and target orbits. These will used to initialize/reset the Kalman filters, and as a backup incase the GPS or CDGPS isn't giving good data and/or the Kalman filter diverges. There will be a state field flag to just use the ground propagators and ignore GPS, which will also be helpful for testing.